cloudinit/config/cc_mounts.py

```diff
@@ -223,6 +223,76 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size, method="auto"):
+    """Size is in MB."""
+    dd, fallocate, auto = ("dd", "fallocate", "auto")
+    if method not in (dd, fallocate, auto):
+        raise ValueError("Unknown method '%s' in create_swapfile" % method)
+
+    tdir = os.path.dirname(fname)
+    util.ensure_dir(tdir)
+
+    fstype = get_fstype(tdir)
+
+    if method == "auto":
+        if fstype in ("xfs", "btrfs"):
+            method = dd
+        else:
+            method = fallocate
+
+    LOG.debug("Creating swapfile in '%s' on fstype '%s' using '%s'",
+              fname, fstype, method)
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    unlink_f(fname)
+
+    msg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+    if method == fallocate:
+        cmd = ['fallocate', '-l', '%dM' % size, fname]
```

Comment:
The bodies of this if statement and the one that follows are almost identical. Should we be refactoring that duplication out to a helper?

Reply:
Agreed as per my last comment. Thanks for the review!